src/executor/simple_partition_pass.h

```diff
@@ -0,0 +1,391 @@
+          outputs.insert({e, count++});
+        }
+      }
+    }
+  });
+  return outputs;
+}
+
+/*!
+ * \brief create new input nodes of the subgraph and plug them
+ * \return the inputs of the subgraph node in the main graph
+*/
+std::vector<nnvm::NodeEntry> GetSubgraphInputs(Graph g, NodeRawPtrSet subgraph_set) {
+  std::vector<nnvm::NodeEntry> inputs;
+  nnvm::NodeEntryMap<nnvm::NodeEntry> entry_map;
+  DFSVisit(g.outputs, [&subgraph_set, &inputs, &entry_map](const nnvm::NodePtr &node){
+    if (subgraph_set.count(node.get())) {
+      for (auto &e : node->inputs) {
+        if (!subgraph_set.count(e.node.get())) {
+          if (entry_map.count(e)) {
+            e = entry_map[e];
+          } else {
+            auto new_node = nnvm::Node::Create();
+            new_node->attrs.name = "input_" + std::to_string(inputs.size());
+            entry_map.insert({e, nnvm::NodeEntry{new_node, 0, 0}});
+            inputs.push_back(e);
+            e.node = new_node;
+            e.index = 0;
+          }
+        }
+      }
+    }
+  });
+  // Fix ordering of w.r.t to topology
+  Graph _g;
+  _g.outputs = g.outputs;
+  const auto &idx = _g.indexed_graph();
+  std::sort(inputs.begin(), inputs.end(),
+      [&idx, &entry_map](const nnvm::NodeEntry lhs, const nnvm::NodeEntry rhs) {
+        return idx.entry_id(entry_map.at(lhs)) < idx.entry_id(entry_map.at(rhs));
+      });
+  return inputs;
+}
+
+std::unordered_map<uint32_t, uint32_t> GetGraphInputsMap(const Graph& g) {
+  std::unordered_map<uint32_t, uint32_t> outputs;
+  auto& idx = g.indexed_graph();
+  outputs.reserve(idx.num_nodes());
+  std::vector<uint32_t> input_nodes = idx.input_nodes();
+  for (size_t i = 0; i < input_nodes.size(); ++i) {
+    outputs[input_nodes[i]] = static_cast<uint32_t>(i);
+  }
+  return outputs;
+}
+
+/*!
+ * \brief helper function to display what nodes are in a specific subset
+ */
+void dispNodesSet(Graph g, NodeRawPtrSet s) {
+  DFSVisit(g.outputs, [&s](const nnvm::NodePtr n){
+    if (s.count(n.get())) {
+      std::cout << "  Y " << n->attrs.name << std::endl;
+    } else {
+      std::cout << "  N " << n->attrs.name << std::endl;
+    }
+  });
+}
+
+/*!
+ * \brief Replace a set of nodes by a subgraph node
+ */
+template<typename FCreateNode>
+Graph ReplaceSubgraphs(Graph&& g, const std::vector<NodeRawPtrSet>& subgraph_sets,
+                       FCreateNode create_subgraph_node) {
+  for (auto subgraph_set : subgraph_sets) {
+    // Create MXNet subgraph
+    Graph subgraph;
+    const auto sub_outputs_in_main = GetSubgraphOutputs(g, subgraph_set);
+    subgraph.outputs.resize(sub_outputs_in_main.size());
+    for (auto p : sub_outputs_in_main) {
+      subgraph.outputs[p.second] = p.first;
+    }
+    // To generate a subgraph an input have to be replace by data node (no op)
+    // and it have to be agnostic to the node from which it's an output
+    // (For exemple even if two inputs are two different outputs from the same node)
```

Comment:
```suggestion
    // (For example, even if two inputs are two different outputs from the same node)
```
Same concern as before... this needs more info...